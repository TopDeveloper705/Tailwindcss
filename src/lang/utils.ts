export function isSpace(str: string): boolean {
  return /\s/.test(str);
}

export function isDigit(str: string): boolean {
  return /^\d+$/.test(str);
}

export function isAlpha(str: string): boolean {
  return /\w/.test(str);
}

export function connect(strings: string[]): string {
  return strings.join(';\n') + ';';
}

export function findGroupEnd(text: string, startIndex = 0): number {
  let level = 1;
  let endBracket = searchFrom(text, '}', startIndex);
  while (endBracket !== -1) {
    const nextBracket = searchFrom(text, '{', startIndex);
    if (endBracket < nextBracket || nextBracket === -1) {
      level--;
      startIndex = endBracket + 1;
      if (level === 0) return endBracket;
    } else {
      level++;
      startIndex = nextBracket + 1;
    }
    endBracket = searchFrom(text, '}', startIndex);
  }
  return -1;
}

export function searchFrom(
  text: string,
  target: string | RegExp,
  startIndex = 0,
  endIndex?: number
): number {
  // search from partial of string
  const subText = text.substring(startIndex, endIndex);
  const relativeIndex = subText.search(target);
  return relativeIndex === -1 ? -1 : startIndex + relativeIndex;
}
