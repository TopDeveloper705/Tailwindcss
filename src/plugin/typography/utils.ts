export function uniq<T>(array: T[]): T[] {
  return Array.from(new Set(array));
}

export function castArray<T>(value: T): unknown[] {
  if (!arguments.length) {
    return [];
  }
  return Array.isArray(value) ? value : [value];
}

export function isUsableColor(color: string, values: string | {[key:string]:string}): boolean {
  return Boolean(values && typeof values === 'object' && color !== 'gray' && values[600]);
}

export const round = (num: number): string =>
  num
    .toFixed(7)
    .replace(/(\.[0-9]+?)0+$/, '$1')
    .replace(/\.0$/, '');

export const rem = (px: number): string => `${round(px / 16)}rem`;

export const em = (px: number, base: number): string => `${round(px / base)}em`;
