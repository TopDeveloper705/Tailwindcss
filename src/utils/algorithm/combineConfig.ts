export default function combineConfig(
  a: { [key: string]: unknown },
  b: { [key: string]: unknown },
  arrayMergeDepth = Infinity,
): { [key: string]: unknown } {
  const output = { ...a };
  for (const [key_of_b, value_of_b] of Object.entries(b)) {
    if (key_of_b in a) {
      const value_of_a = a[key_of_b];
      if (value_of_a !== value_of_b) {
        if (value_of_b !== null && (value_of_b as string).constructor !== Object) {
          if (arrayMergeDepth > 0 && Array.isArray(value_of_a) && Array.isArray(value_of_b)) {
            output[key_of_b] = [...value_of_a, ...value_of_b];
          } else {
            output[key_of_b] = value_of_b;
          }
        } else if (value_of_a !== null && (value_of_a as { [key: string]: unknown }).constructor === Object) {
          output[key_of_b] = combineConfig(
            value_of_a as { [key: string]: unknown },
            value_of_b as { [key: string]: unknown },
            arrayMergeDepth - 1
          );
        } else if (Array.isArray(value_of_a)){
          output[key_of_b] = [
            ...value_of_a,
            ...Array.isArray(value_of_b) ? value_of_b : [value_of_b],
          ];
        } else {
          output[key_of_b] = {
            DEFAULT: value_of_a,
            ...value_of_b as { [key: string]: unknown },
          };
        }
      }
    } else {
      output[key_of_b] = value_of_b;
    }
  }
  return output;
}
