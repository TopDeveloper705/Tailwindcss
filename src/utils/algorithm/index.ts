export { default as sortSelector } from './sortSelector';
export { default as sortMediaQuery } from './sortMediaQuery';
export { default as compileStyleSheet } from './compileStyleSheet';
export { default as cssEscape } from './cssEscape';
export { default as combineConfig } from './combineConfig';
export { default as diffConfig } from './diffConfig';
