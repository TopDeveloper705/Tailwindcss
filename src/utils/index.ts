export * from "./tools";
export * from "./helpers";
