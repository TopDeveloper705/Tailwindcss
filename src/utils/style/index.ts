export { Property, InlineAtRule, Style, GlobalStyle, Keyframes } from './base';
export { StyleSheet } from './sheet';
export * from './prefixer';
