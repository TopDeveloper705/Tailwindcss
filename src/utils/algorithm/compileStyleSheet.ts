import sortMediaQuery from './sortMediaQuery';
import { sortMeta } from './sortSelector';
import { Keyframes, Container, Style } from '../style/base';
import { wrapit, hash, isSpace } from '../../utils/tools';
import type { AnyObject } from '../../interfaces';

export function combineObject(
  a: { [key: string]: unknown },
  b: { [key: string]: unknown }
): { [key: string]: unknown } {
  const output = { ...a };
  for (const [key_of_b, value_of_b] of Object.entries(b)) {
    if (key_of_b in a) {
      const value_of_a = a[key_of_b];
      if (value_of_a !== value_of_b) {
        if (
          value_of_b !== null &&
          (value_of_b as string).constructor !== Object
        ) {
          output[key_of_b] = [value_of_a, value_of_b];
        } else if (
          value_of_a !== null &&
          (value_of_a as { [key: string]: unknown }).constructor === Object
        ) {
          output[key_of_b] = combineObject(
            value_of_a as { [key: string]: unknown },
            value_of_b as { [key: string]: unknown }
          );
        } else {
          output[key_of_b] = [
            value_of_a,
            combineObject(
              value_of_a as { [key: string]: unknown },
              value_of_b as { [key: string]: unknown }
            ),
          ];
        }
      }
    } else {
      output[key_of_b] = value_of_b;
    }
  }
  return output;
}

export function deepList<T>(
  list: string[],
  value?: T
): { [key: string]: unknown } {
  const key = list.pop();
  const current = value ? value : {};
  if (!key) return current;
  const dict: { [key: string]: AnyObject } = {};
  dict[key] = current;
  return deepList(list, dict);
}

export function handleNest(item: unknown): unknown[] {
  let output: unknown[] = [];
  if (Array.isArray(item)) {
    item.forEach((i) => {
      output = [...output, ...handleNest(i)];
    });
  } else {
    if ((item as { build?: () => '' }).build)
      output.push((item as { build: () => '' }).build());
  }
  return output;
}

export function buildMap(obj: unknown, minify = false, prefixer = true): string {
  let output: unknown[] = [];
  if (Array.isArray(obj)) {
    obj.forEach((item) => {
      if (item.constructor === Object) {
        output.push(buildMap(item));
      } else if (Array.isArray(item)) {
        output = [...output, ...handleNest(item)];
      } else {
        if (item.build) output.push(item.build(minify));
      }
    });
  } else if (obj && typeof obj === 'object') {
    for (const [key, value] of Object.entries(obj)) {
      const _gstyle = (v: string) =>
        (minify ? key.replace(/\n/g, '') : key + ' ') +
        wrapit(v, undefined, undefined, undefined, minify);
      if (value instanceof Style) {
        output.push(_gstyle(value.build(minify, prefixer)));
      } else if (value && typeof value === 'object') {
        output.push(_gstyle(buildMap(value, minify, prefixer)));
      }
    }
  }
  return output.join(minify ? '' : '\n');
}

export function combineSelector(styleList: Style[]): Style[] {
  const styleMap: { [key: string]: Style } = {};
  const passed: Style[] = [];
  styleList.forEach((style) => {
    const rule = style.rule;
    if (rule) {
      const hashValue = hash(rule);
      if (hashValue in styleMap) {
        styleMap[hashValue] = styleMap[hashValue].extend(style, true);
      } else {
        styleMap[hashValue] = style;
      }
    } else {
      passed.push(style);
    }
  });
  return [...passed, ...Object.values(styleMap).map((style) => style.clean())];
}

export function buildAtRule(styleList: Style[], minify = false, prefixer = true, reverse = false): string {
  const ruleMap = styleList
    .map(i => {
      const list = [
        ...reverse ? (i.atRules ?? []).sort(sortMediaQuery) : (i.atRules ?? []).sort(sortMediaQuery).reverse(),
        i.rule,
      ];
      const style = new Style(undefined, i.property, i.important);
      style.meta = i.meta;
      i.wrapProperties && i.wrapProperties.forEach(wrap => style.wrapProperty(wrap));
      return deepList(list, style);
    })
    .sort((a, b) => {
      const akey = Object.keys(a)[0];
      const bkey = Object.keys(b)[0];
      return sortMediaQuery(akey, bkey);
    });
  return buildMap(ruleMap.reduce((previousValue, currentValue) => combineObject(previousValue, currentValue), {}), minify, prefixer);
}

export default function compileStyleSheet(
  styleList: Style[],
  minify = false,
  prefixer = true,
): string {
  const head = combineSelector(styleList.filter((i) => !(i.selector && i.atRules) && !(i instanceof Container)).sort(sortMeta)).map((i) => i.build(minify, prefixer)).join(minify ? '' : '\n');
  const containers: {[key:string]: Container[]} = {};
  styleList.filter(i => i instanceof Container).forEach(i => {
    if (i.selector && i.selector in containers) {
      containers[i.selector].push(i);
    } else if (i.selector) {
      containers[i.selector] = [i];
    }
  });
  const topStyles = Object.values(containers).map(i => buildAtRule(i, minify, prefixer, true));
  const keyframes = buildAtRule(styleList.filter(i => i instanceof Keyframes), minify, prefixer);
  if (keyframes) topStyles.unshift(keyframes);
  const top = topStyles.join(minify? '' : '\n');
  const body = buildAtRule(styleList.filter((i) => i.selector && i.atRules && !(i instanceof Keyframes || i instanceof Container)).sort(sortMeta), minify, prefixer);
  return minify
    ? (top + head + body).replace(/;\}/g, '}')
    : [top, head, body].filter((i) => !isSpace(i)).join('\n');
}
