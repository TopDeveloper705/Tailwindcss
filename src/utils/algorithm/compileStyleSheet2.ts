import { Style } from '../style/base';
import { wrapit, deepCopy } from '../../utils/tools';

function _buildAtrule(atrule:string, children: Style[], minify = false, prefixer = true) {
  return `${ atrule }${ minify ? '' : ' ' }${ wrapit(_buildStyleList(children, minify, prefixer), undefined, undefined, undefined, minify) }`;
}

function _buildStyleList(styleList: Style[], minify = false, prefixer = true): string {
  let currentAtrule: string | undefined;
  let currentStyle: Style | undefined;
  let styleStack: Style[] = [];
  const output: string[] = [];
  for (const style of styleList) {
    if (style.isAtrule) {
      if (currentStyle) {
        output.push(currentStyle.clean().build(minify, prefixer));
        currentStyle = undefined;
      }
      const newAtrule = (style.atRules as string[]).pop();
      if (currentAtrule) {
        if (currentAtrule === newAtrule && newAtrule !== '@font-face') { // @font-face shouldn't been combined
          styleStack.push(style);
        } else {
          output.push(_buildAtrule(currentAtrule, styleStack, minify, prefixer));
          currentAtrule = newAtrule;
          styleStack = [ style ];
        }
      } else {
        currentAtrule = newAtrule;
        styleStack = [ style ];
      }
    } else {
      if (currentAtrule) {
        output.push(_buildAtrule(currentAtrule, styleStack, minify, prefixer));
        currentAtrule = undefined;
        styleStack = [];
      }
      if (currentStyle) {
        if (style.rule === currentStyle.rule) {
          if (style.important) style.property.forEach(p => p.important = true);
          if (style.wrapProperties) style.property.forEach(p => style.wrapProperties?.forEach(wrap => p.name = Array.isArray(p.name) ? p.name.map(i => wrap(i)) : wrap(p.name)));
          currentStyle.add(style.property);
        } else {
          output.push(currentStyle.clean().build(minify, prefixer));
          currentStyle = style;
        }
      } else {
        currentStyle = style;
      }
    }
  }
  if (currentAtrule) output.push(_buildAtrule(currentAtrule, styleStack, minify, prefixer));
  if (currentStyle) output.push(currentStyle.clean().build(minify, prefixer));
  return output.join(minify ? '' : '\n');
}

export default function (styleList: Style[], minify = false, prefixer = true): string {
  return _buildStyleList(deepCopy(styleList), minify, prefixer);
}
