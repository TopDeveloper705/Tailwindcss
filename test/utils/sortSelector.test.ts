import { Style } from '../../src/utils/style';
import { sortSelector } from '../../src/utils/algorithm';

describe('sort selector', () => {
  it('should sort correctly', () => {
    const styles = [
      new Style(':root'),
      new Style('::moz-focus-inner'),
      new Style('::-webkit-inner-spin-button'),
      new Style('*'),
      new Style('.test'),
      new Style('a'),
      new Style('#id'),
      new Style('body'),
      new Style('.bg-red-500'),
      new Style('.p-4'),
      new Style('.pt-4'),
      new Style('.p-2'),
    ];
    const selectors = styles.sort(sortSelector).map((i) => i.selector);
    expect(selectors).toMatchSnapshot('selector', __filename);
  });

  it('should sort transform correctly', () => {
    const styles = [
      new Style('.-translate-x-1/2'),
      new Style('.-translate-y-1/2'),
      new Style('.origin-top'),
      new Style('.transform'),
      new Style('.scale-50'),
      new Style('.rotate-5'),
      new Style('.-skew-x-4'),
    ];
    const selectors = styles.sort(sortSelector).map((i) => i.selector);
    expect(selectors).toMatchSnapshot('selector', __filename);
  });
});
