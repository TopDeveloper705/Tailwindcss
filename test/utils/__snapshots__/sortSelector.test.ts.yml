Tools / should sort correctly / selector / 0: |-
  [
    "*",
    ":root",
    "::moz-focus-inner",
    "::-webkit-inner-spin-button",
    "a",
    "body",
    ".bg-red-500",
    ".p-4",
    ".p-2",
    ".pt-4",
    ".test",
    "#id"
  ]
Tools / should sort transform correctly / selector / 0: |-
  [
    ".transform",
    ".origin-top",
    ".scale-50",
    ".rotate-5",
    ".-translate-x-1/2",
    ".-translate-y-1/2",
    ".-skew-x-4"
  ]
